### ISSUE TITLE
"do / else" proposal

### ISSUE THREAD

**Issue by certik** [j3-fortran/fortran_proposals#342]
https://fortran-lang.discourse.group/t/for-else-or-do-else-in-fortran/8696

Example:
```fortran
condition = .false.
do i = 1, 10
    .......
    if (condition) exit
else
    ! This block only executes if the do loop doesn't exit.
    error stop "Condition not met"
enddo
```

Similar to Python's "for else": https://docs.python.org/3/tutorial/controlflow.html#else-clauses-on-loops

**Comment by everythingfunctional** [comment]
Just to be clear, the above is equivalent to the below?

```Fortran
condition = .false.
do i = 1, 10
    ...
    if (condition) exit
end do
if (i > 10) then
    ! This block only executes if the do loop doesn't exit.
    error stop "Condition not met"
end if
```

**Comment by certik** [comment]
Yes. I updated the above description with an example. (I don't like using loop variables after the loop, but that's a separate issue: is it even specified by the standard?)

**Comment by everythingfunctional** [comment]
> is it even specified by the standard?

Yes, although it's not immediately obvious. From 11.1.7.4.3 The execution cycle

> The DO variable, if any, is incremented by the value of the incrementation parameter m3

and from 11.1.7.4.5 Loop termination

> When a DO construct becomes inactive, the DO variable, if any, of the DO construct retains its last defined value.

So if a do loop completes all its iteration, the standard says it has a value > the ending value.

**Comment by certik** [comment]
@everythingfunctional thanks for the clarification!

**Comment by klausler** [comment]
Looks like syntactic vinegar for the existing and less mysterious

```
search: block
  do j = 1, 10
    ...
    if (condition) exit search
  end do
  error stop 'nope'
end block
```

**Comment by chuckyvt** [comment]
Try fully sketch this out, I think the following example would also be included.

```
condition = .false.
i = 1
do while (i < 11 )
    .......
    if (condition) exit
    i = i + 1
else
    ! This block only executes if the do loop doesn't exit.
    error stop "Condition not met"
enddo
```

Not exactly sure how the below case should be handled.  Python explicitly ties it to a break statement, so perhaps it doen't execute.    
```
condition = .false.
do 
    .......
    if (condition) go to 10
10 else
    ! This doesn't execute since there is no exit statement?  
    error stop "Condition not met"
enddo
```

**Comment by chuckyvt** [comment]
I think current approaches to catch failures can be lumped into two general groups, along with what I believe are limitations of those approaches.  I believe these limitations would mostly be addressed with a do / else approach.       

GO TO / Exit Block:  Use either a GO TO or exit block statement to jump forward and skip the error condition code.   This is likely the most common approach.  The downsides is jump statements can become difficult to follow and breaks up the flow of the code. 

Check loop variable after loop completes:  It's not obvious at first glance that the if statement is conditionally executed based on the loop exit state.  And since the if block isn't explicitly tied to the do loop,  the do conditional statement can get out of sync with the if conditional statement if the programmer isn't careful. 

For an interesting discussion of this feature in Python, see below link . 
https://stackoverflow.com/questions/9979970/why-does-python-use-else-after-for-and-while-loops
